apis/Google.Cloud.Firestore/docs/datamodel.md

```diff
@@ -147,14 +147,99 @@ document data to a dictionary representation.
 
 The equivalent city code using dictionaries would look like this:
 
-[!code-cs[](obj/snippets/Google.Cloud.Firestore.DataModel.txt#DictionaryUsage)]
+{{sample:DataModel.DictionaryUsage}}
 
 ## Mapping with anonymous types
 
 Anonymous types can be used for serialization, but not deserialization. They are particularly useful to specify partial updates, or
 to populate data which isn't then read within the same codebase.
 
-[!code-cs[](obj/snippets/Google.Cloud.Firestore.DataModel.txt#AnonymousTypeUsage)]
+{{sample:DataModel.AnonymousTypeUsage}}
+
+## Custom converters
+
+If the built-in conversions aren't flexible enough for your needs,
+you can create a custom converter implementing
+`IFirestoreConverter<T>`:
+
+```csharp
+public interface IFirestoreConverter<T>
+{
+    object ToFirestore(T value);
+    T FromFirestore(object value);
+}
+```
+
+The `ToFirestore` method should convert the `T` value into a
+suitable format to be stored. This can use any of the conversions
+described above. For example, if the method returns an `int`, the
+value will be stored as a 64-bit integer in the same way as an `int`
+property in an attributed type, or an `int` value within a dictionary.
+
+The `FromFirestore` method receives the deserialized value using the
+default mapping, as shown in the earlier table.
+
+### Applying a converter to a type
+
+Once you've created a converter, you can either apply it to a type
+or an individual attributed property. In each case, the
+`ConverterType` property is specified in the attribute.
+
+For example, suppose you wish to create several ID classes or
+structs, each containing a string of the underlying ID. This is a
+technique sometimes used to effectively make type-safe identifiers.
+You could create a converter for each class, and apply the
+`[FirestoreData]` attribute to each class, specifying the
+corresponding converter. This would allow each ID to be stored as
+just the string, rather than as a map containing a single element with
+a string value. The following example demonstrates this with a
+`PlayerId` class. If another attributed class (e.g. a `Game`) had a
+`PlayerId` property, the converter would be used automatically.
+
+{{sample:DataModel.CustomConverterType}}
+
+### Applying a converter to a property
+
+Sometimes you may want to perform custom conversions for types that
+you can't apply attributes to, or you may want different conversions
+in different situations. In that case, you can implement the
+converter in the same way, but apply it selectively using the
+`CustomConverter` property on the `[FirestoreProperty]` attribute
```

Comment:
using the `ConverterType` property on the...

Reply:
Whoops, thanks - will fix.